src/test/kotlin/datamaintain/SorterTest.kt

```diff
@@ -0,0 +1,48 @@
+package datamaintain
+
+import org.junit.jupiter.api.Test
+import strikt.api.expectThat
+import strikt.assertions.first
+import strikt.assertions.get
+import strikt.assertions.isEqualTo
+import strikt.assertions.last
+import java.nio.file.Paths
+
+internal class SorterTest {
+    private val sorter: Sorter = Sorter(Config(Paths.get(""), ""))
+
```

Comment:
Case sensitivity is a topic on its whole regarding file names : I remember that on (old) version of Windows (not sure if it's still the case) file names matching were case insensitive (`script1.js` was considered the same as `ScRiPt1.JS`)

Not sure what to think about it (if we should enforce case sensitivity, or consider file name as case insensitive just because of windows)

About the number mixing, I don't think that we should allow both `1-script2.js` and `script2.js` at the same time.
I can recall that at some point, we talked about a regex pattern allowing to "extract" ordering number from file name
=> If we use such regex to consider that the number "is just before the file extension", then number for `1-script2.js` will be `2` (the `1-` prefix will be ignored, due to the regex)

Reply:
Hello,

About case sensitivity, I think we can be simple : 'flattening case' with lowercase for instance and then sort.

About number mixing: yes about the regexp to extract a prefix but it is not really my point (and about the prefix, in a next version it will be the value used to sort). 
I don't really understand why we should not allow to mix both (1-script2.js and script2.js). I am agree to say it is ugly and it is not the way we will recommand, but for me we should be dumb and apply an alphabetical ordering. So 1-script2.js should be before script2.js.

But my samples was bad because my initial point was to ensure script1.js, script11.js and script111.js are correctly sorted.

Reply:
I do not think case sensitivity would be important for this feature. I mean what difference would be for a developer if their script's name is 'script1' or 'Script1'? I do not think anyone would expect a difference in the order just because of that uppercase. The only problem that it may cause is when two scripts have the same name excluding the case but well since having two scripts named the same is not something we'd like to encourage... 

Just to make sure that case does not matter, I would do as @driccio suggested: flattening case before sorting.

I do not understand either why naming a script 'script1' would be a problem. What matters anyway is the number before the script's name, that '1' would be just considered as part of the script's name. I would expect the regexp allowing us to extract the number to look for that number before a specific character, like - and though it would only evaluate the '1' in 'script1' as that number because that script would be badly named and do not contain the separator that was chosen. In a script '1-script1', there would be no problem. 

As for numbers in scripts names, having a test that would ensure that scripts named script1.js, script11.js and script111.js are correctly sorted sounds like a good idea to me :) That was kind of what we wanted to do with the second test we wrote but we missed the point when putting numbers at the beginning of our scripts names.

Reply:
> Just to make sure that case does not matter, I would do as @driccio suggested: flattening case before sorting.

I'm in favour of case insensitivity too. 
To go one step further, I would even *store* script names as upper-cased (or lower-cased, it depends of taste). I guess that's what you were proposing @driccio ?
> About case sensitivity, I think we can be simple : 'flattening case' with lowercase for instance and then sort.

About the number mixing, I think we don't understand ourselves.
My point was to say "as soon as we have a regexp aimed at extracting a `$scriptIdentifier`", then we should base the sorting upon this `$scriptIdentifier`, not the raw script name (otherwise, this `$scriptIdentifier` will be useless).
If we don't have any regexp aimed at extracting a `$scriptIdentifier` then the `$scriptIdentifier` _is_ the filename (implicit : we sort alphabetically based on filename)

Makes sense ?

Reply:
> My point was to say "as soon as we have a regexp aimed at extracting a `$scriptIdentifier`, then we should base the sorting upon this `$scriptIdentifier`, not the raw script name (otherwise, this `$scriptIdentifier` will be useless).
> If we don't have any regexp aimed at extracting a `$scriptIdentifier` then the `$scriptIdentifier` _is_ the filename/ffilepath (implicit : we sort alphabetically based on filename/filepath)
> 
> Makes sense ?

Oh yes, sure. I agree with that.

Reply:
Hello,

> To go one step further, I would even _store_ script names as upper-cased (or lower-cased, it depends of taste). I guess that's what you were proposing @driccio ?
I didn't speak about storage. I prefer to store the real name of the file but considering it lower-cased name in the algo.


> About the number mixing, I think we don't understand ourselves.
> My point was to say "as soon as we have a regexp aimed at extracting a `$scriptIdentifier`, then we should base the sorting upon this `$scriptIdentifier`, not the raw script name (otherwise, this `$scriptIdentifier` will be useless).
Yes of course, but I don't read anything telling it would not be the case. The prefix we can extract is a substring of the file name, but it can also be the filename itself. So for me there is no reason to make distinction between the extracted prefix or the filename. The algorithm will be the same to sort this string. I guess we are all ok with that.

Reply:
> So for me there is no reason to make distinction between the extracted prefix or the filename. The algorithm will be the same to sort this string. I guess we are all ok with that.

The important question here is "what is the input type for the sorting algorithm" ?

- Should it be the `filepath` only ?
- Should it be what I call the `$scriptIdentifier` (which would be the result of a `ScriptIdentifierExtractor`, configurable, getting the script's `filepath` as input) only and which could be, sometimes (when default `ScriptIdentifierExtractor` is used), valued to the script's `filepath`
- Should it be a `Script` object having a lot of informations embedded into it (`filepath`, `scriptIdentifier`, `metadata` etc...)

Thinking about it, I would be more in favour of the 3rd option which allows a lot of flexibility in the sorting algorithm.

And by default, the sorting would be based on `Script.filepath` attribute.

Reply:
Hello Fred,

For me it is also the 3rd option because we can not exclude to have other sorting strategies.

> And by default, the sorting would be based on `Script.filepath` attribute.
Sorry but I disagree. By default  the sorting would be based on `Script.scriptIdentifier`. The `scriptIdentifier` can be equal to the filename (not filepath) if no regexp is set to extract a prefix. If this regexp exists then the `scriptIdentifier` will be the extracted prefix.

Reply:
Agree, I used a shortcut to talk about `Script.filepath` instead of the `scriptIdentifier`, but basically, if we consider we rely on the default behaviour (no `ScriptIdentifierExtractor`), then this should be the same.

That being said, I don't understand why you would take `filename` in consideration rather than the whole `filepath` ?

We don't want `directory1/script5` to be executed *after* `directory2/script3`, and I assume this is something obvious.

I feel it implicitely means that you're going to 1/ group `Script`s by directory, then 2/ use sort algorithm to sort `Script`s belonging to the same directory
... but it doesn't tell how to sort directories altogether.

Using `filepath` solves this issue.